feat: Add event type filter for subscriptions

david-marconis opened the pull request:
## Description of changes
Add support for event type filtering to the Event Grid emulator, allowing users to specify which event types are delivered to specific subscriptions. The configuration is based on the event filtering offered by Azure event grid:
https://learn.microsoft.com/en-us/azure/event-grid/event-filtering#event-type-filtering

* Add config logging of option changes for filters
* Clean up Equals and HashCode implementations for ToipicOptions
* Add abstract method FilterEvents to BaseEventHttpContextHandler for filtering the events.

## Breaking changes
The output of the config load/reload is slightly different.
I intentionally added a new config item instead of modifying the existing Topics to not break backwards compatibility.

## Additional checks

- [x] Updated the documentation of the project to reflect the changes
- [x] Added new tests that cover the code changes

Gcaya commented:
Hello, thanks for your contribution. I'm taking a look at it!

Gcaya commented on README.md:
````diff
@@ -55,6 +55,31 @@ In the example for push delivery, we have two topics, `topic1` and `topic2`. If
+```json
+{
+  "Topics": {
+    "topicfoobar": [
````
Can you maybe add a push subscription as well? Just to show that filtering works for both models

david-marconis commented:
I added a test to test the push-based filtering as well.

In doing so I noticed a bug where if you had an URL without a slash in the path it would be added because of the corrected subscribers filtering in `PostConfigureTopicOptionsCorrector`. I added a similar correction to the filters when the URL of the subscription was corrected. It will also detect invalid filters if there is a filter pointing to a non-existent subscription.

I also updated the documentation to include the push-based filtering.

I hope you find the changes appropriate.

Gcaya commented:
Looks good to me! Gonna close this PR and merge the changes in this one https://github.com/gsoft-inc/wl-eventgrid-emulator/pull/111

Don't worry all your commits have been kept, it's just our processes for merging external contributions are not fully defined yet.
